lib/bundler/match_platform.rb

```diff
@@ -15,7 +15,10 @@ def self.platforms_match?(gemspec_platform, local_platform)
       return true if Gem::Platform::RUBY == gemspec_platform
       return true if local_platform == gemspec_platform
       gemspec_platform = Gem::Platform.new(gemspec_platform)
-      return true if GemHelpers.generic(gemspec_platform) === local_platform
+      generic_gemspec_platform = GemHelpers.generic(gemspec_platform)
+      unless generic_gemspec_platform == Gem::Platform::RUBY
```

Comment:
This isn't needed for platforms that are listed in `Bundler::GemHelpers::GENERICS`. Because these platforms are resolved to a `Gem::Platform` object. Platforms that aren't listed such as `x86_64-linux` are resolved to `Gem::Platform::RUBY`. See `Bundler::GemHelpers.generic` for details: https://github.com/rubygems/bundler/blob/master/lib/bundler/gem_helpers.rb#L21

If `Gem::Platform::RUBY` is returned, it's matched for `local_platform` is `Gem::Platform::RUBY`. It means that this method returns `true` for `x86_64-linux` gem on `ruby` platform. But a platform specific gem such as a gem for `x86_64-linux` doesn't work all platform (= `ruby` platform).

So I added this condition.

Reply:
Umm... I think that this is a bug but some existing examples are failed with this change:

https://travis-ci.org/rubygems/bundler/jobs/638212688#L1303

```text
rspec ./spec/cache/gems_spec.rb:236 # bundle cache when previously cached doesn't remove gems that are for another platform
rspec ./spec/commands/install_spec.rb:220 # bundle install with gem sources the simple case with a gem that installs multiple platforms installs gems for the local platform as first choice
rspec ./spec/commands/update_spec.rb:39 # bundle update with --all updates the entire bundle
rspec ./spec/commands/update_spec.rb:49 # bundle update with --all doesn't delete the Gemfile.lock file if something goes wrong
rspec ./spec/commands/update_spec.rb:104 # bundle update with a top level dependency unlocks all child dependencies that are unrelated to other locked dependencies
rspec ./spec/commands/update_spec.rb:115 # bundle update with an unknown dependency should inform the user
rspec ./spec/commands/update_spec.rb:119 # bundle update with an unknown dependency should suggest alternatives
rspec ./spec/commands/update_spec.rb:126 # bundle update with a child dependency should update the child dependency
rspec ./spec/commands/update_spec.rb:299 # bundle update in a frozen bundle should suggest different command when frozen is set globally
rspec ./spec/runtime/platform_spec.rb:60 # Bundler.setup with multi platform stuff will add the resolve for the current platform
```

So I've reverted this change for now. And the new example uses `simulate_platform mingw`. It means that the new example is failed without `simulate_platform` on `x86_64-linux`.